File: Solutions/23.py
# -*- coding: utf-8 -*-


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def main():
    lists = []
    from queue import PriorityQueue

    que = PriorityQueue()
    for i in range(len(lists)):
        if lists[i]:
            que.put((lists[i].val, i))

    head = cur = ListNode(0)
    while not que.empty():
        v, idx = que.get()
        node = lists[idx]
        cur.next = node
        cur = cur.next
        if node.next:
            lists[idx] = lists[idx].next
            que.put((node.next.val, idx))
    return head.next

File: Solutions/23.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <queue>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

struct cmp{
    bool operator()(ListNode *a,ListNode *b){
        return a->val > b->val;
    }
};

ListNode* mergeKLists(vector<ListNode*>& lists) {
    int k = lists.size();
    priority_queue<ListNode*, vector<ListNode*>, cmp> que;
    for(int i=0;i<k;++i) if(lists[i])que.push(lists[i]);

    ListNode* head = new ListNode(0);
    ListNode* cur = head;
    while(!que.empty()){
        ListNode* tmp = que.top(); que.pop();
        cur->next = tmp;
        cur = cur->next;
        tmp = tmp->next;
        if(tmp) que.push(tmp);
    }
    return head->next;
}
